src/libmongoc/doc/mongoc_client_get_timeout_ms.rst

```diff
@@ -0,0 +1,24 @@
+:man_page: mongoc_client_get_timeout_ms
+
+mongoc_client_get_timeout_ms()
+==============================
+
+Synopsis
+--------
+
+.. code-block:: c
+
+  int64_t
+  mongoc_client_get_timeout_ms (const mongoc_client_t *client)
+
+Returns the client-side operation timeout value for this object. If ``client`` does not have a timeout currently set, this method will return the default timeout, which is 0, or infinite timeout.
```

Comment:
Based on the `Backwards Breaking Considerations` section of the spec, I think unset will mean "keep the existing behavior". Otherwise, when we introduce `timeoutMS` support, it could silently override existing default timeout.

E.g. the URI "mongodb://localhost:27017/?sockettimeoutms=1000" may suddenly change to mean no socket timeout (with a deprecation warning).

This might just be a doc change:
```suggestion
Returns the client-side operation timeout value for this object. If ``client`` does not have a timeout currently set, this method will return 0.
```

`0` will have a double meaning here. I am ok with that since callers can determine with `mongoc_client_timeout_is_set`. If we want to avoid that, perhaps this could return an `const int64_t*`, which is NULL on unset. Though I'm fine with this as-is.

Reply:
Would the returned `int64_t*` point to the `int64_t` _within_ the internal timeout struct?

I suppose this is consistent with the RC/RP/WC getters, which return the pointer _within_ the client struct. In that case, the pointers are valid for either the lifetime of the client struct or until a new RC/RP/WC is assigned, whichever comes first.

It seems a bit odd to do for an integer type, though. AFAIK, the CSOT spec doesn't prohibit you from returning a negative value (i.e. -1) to communicate that it's unset -- I don't imagine that'd be different than other languages returning a nullable integer. You could also define a public constant (e.g. `TIMEOUTMS_UNSET` or `TIMEOUTMS_DEFAULT`), as is done for `MONGOC_WRITE_CONCERN_W_DEFAULT`. I think the only difference from WC would be that you wouldn't be able to specify this constant via the setter methods -- but that seems fine.

Just my $.02.

Reply:
Yes, I was thinking it would return a pointer to within the struct.

A negative `TIMEOUTMS_UNSET` sentinel seems like a good idea too. I think it is unlikely a negative `timeoutMS` will ever mean something.

FWIW I am also ok with the current proposed API with two functions.

Reply:
I suppose if you opt for defining `TIMEOUTMS_UNSET` as a negative, you could get by with removing both the `is_set` function _and_ the struct. The public APIs that set `timeoutMS` (e.g. URI, setter methods, method options) could just require a non-negative value (I assume they may already) and fall back to -1 or inherit as the default behavior accordingly.

Seems out of scope for this PR. Do you want to take it offline for future consideration?

Reply:
I've changed the documentation as @kevinAlbs suggested.

I agree with @jmikola that it would be simpler to use -1 as a special value to mean `TIMEOUTMS_UNSET`.  In the interest of not flip-flopping too much on this code review, I'd like consensus before making that change (because it will change a lot of the code in this review).  So, thoughts, @kevinAlbs and @bazile-clyde ?  If we use -1 as a special value, I will remove the `is_set` getter methods.

Reply:
Defining a `TIMEOUTMS_UNSET` as `-1` sounds good to me!

Reply:
Sounds good to me too.

Reply:
![](https://media.giphy.com/media/3o6MbdoAQ6faAW5xDi/giphy.gif)